Add RunTemplate and preliminary support for TaskGroup

timcharper opened the pull request.

Commit:
WIP

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/Ids.scala:
```diff
@@ -6,6 +6,16 @@ package com.mesosphere.usi.core.models
 case class PodId(value: String)
 
+/**
+  * Identifier for a pod's task; does not include podId
```
Could you give an example?

timcharper replied:
Added docs

timcharper replied:
And tests

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/RunTemplateLike.scala:
```diff
@@ -0,0 +1,134 @@
+    }
+    executorResources.foreach { b += _ }
+    b.result()
+  }
```
`allResources = taskResourceRequirements.valueIterator.toVector` should do exactly that.

timcharper replied:
This has been replaced with a Map, and, please see this included the executorResources also

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/RunTemplateLike.scala:
```diff
@@ -0,0 +1,134 @@
+  val taskResourceRequirements: Map[PartialTaskId, Seq[ResourceRequirement]]
+  val executorResources: Seq[ResourceRequirement]
+  val role: String
+  private[usi] def buildOperation(
```
What do you think about
```suggestion
  private[usi] def buildLaunchOperation(
```

timcharper replied:
Sure

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/RunTemplateLike.scala:
```diff
@@ -0,0 +1,134 @@
+
+  override private[usi] def buildOperation(
+      matchedOffer: Mesos.Offer,
+      taskIdStrategy: CurriedPodTaskIdStrategy,
```
If I understand correctly the user can define how tasks are named. Should we allow this flexibility?

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/Ids.scala:
```diff
@@ -6,6 +6,16 @@ package com.mesosphere.usi.core.models
+  * Identifier for a pod's task; does not include podId
+  * @param value
+  */
+case class PartialTaskId(value: String)
```
What is the relation between a `TaskId` and a `PartialTaskId`?

timcharper replied:
Clarified this by renaming to TaskName.

Also, docs.

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/RunTemplateLike.scala:
```diff
@@ -0,0 +1,134 @@
+    : Either[Mesos.Offer.Operation.Launch, Mesos.Offer.Operation.LaunchGroup]
+}
+
+class LegacyLaunchBuilder(val role: String, taskBuilder: TaskBuilderLike) extends RunTemplateLike {
```
This is a little confusing to me. Maybe it's the naming but why would a builder extend a template?

jeschkies replied:
How about making the build a field of the template?

timcharper replied:
Yeah, WIP PR did not include attention to naming :) I've renamed it. I think this is an ok use of inheritance but am happy to hear your ideas on how composition would be better here.

jeschkies reviewed (commented):
Thanks. I have some open questions.

Commit:
Tests pass

Commit:
Refactoring

Fix unit test failure

scala format

Commit:
Only use `--cgroups_root` flag when starting Mesos agents if OS is Linux

The flag --cgroups_root is not supported on Mac OS and causes integration tests
to fail.

Commit:
Rename PartialTaskId to TaskName

Also, add unit tests for TaskName

Commit:
buildOperation => buildLaunchOperation

Commit:
Rename LaunchGroupBuilder to LaunchGroupRunTemplate

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/linting/Linting.scala:
```diff
@@ -0,0 +1,47 @@
+import com.mesosphere.usi.core.models.template.RunTemplate
+import com.mesosphere.usi.core.models.template.RunTemplate.KeyedResourceRequirement
+
+class Linting {
```
This is not a blocker IMHO but what to you think about using Scalactic's [Or and Every](http://www.scalactic.org/user_guide/OrAndEvery)? We probably should decide for a validation pattern earlier than later.

timcharper replied:
As mentioned, this code was moved our of MesosEventLogic. I think I'd prefer to not do a lot with it right now, other than to move it, since much of it doesn't apply in the RunTemplate world.

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/Ids.scala:
```diff
@@ -6,6 +6,17 @@ package com.mesosphere.usi.core.models
 case class PodId(value: String)
 
+/**
+  * Name for a pod's task. The TaskName and PodId are combined via PodTaskIdStrategy to generate a TaskId.
```
Nice. So in Marathon world we might have `myGroup_myApp.marathon-b6ff5fa5-7714-11e7-a55c-5ecf1c4671f6._app.1`.
*  `myGroup_myApp.marathon-b6ff5fa5-7714-11e7-a55c-5ecf1c4671f6` is the instance id aka pod id in USI
* `_app.1` is the task name
* both combined are the task id

Did I get this right?

timcharper replied:
Yeah, that's right.

Also, the PodTaskIdStrategy will be pluggable so that frameworks can customize how this maps

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/PodTaskIdStrategy.scala:
```diff
@@ -0,0 +1,39 @@
+          logger.error(
+            s"taskName ${taskName.value} for podId ${podId.value} contains a '.', which is the delimiter character; this results in an unparseable taskId.")
+        }
+        TaskId(podId.value + "." + taskName.value)
```
This is probably too much for now but I think `TaskId` should be a trait so that we can define `case class DefaultTaskId(podId: PodId, taskName: TaskName)`. It's always convenient to extract them again without parsing.

timcharper replied:
Ah, interesting point!

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/RunningPodSpec.scala:
```diff
@@ -31,45 +31,4 @@ case class RunningPodSpec(id: PodId, runSpec: RunTemplate) extends PodSpec {
-    uniqueRangeRequirements ++ scalarRequirements
-  }
-}
+object RunningPodSpec {}
```
Where is this used?

timcharper replied:
Not sure if IntelliJ added this for me or if I was going to add a method and then quit :) I'll clean up

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/TaskBuilder.scala:
```diff
@@ -0,0 +1,12 @@
+import com.mesosphere.usi.core.models.resources.ResourceRequirement
+import org.apache.mesos.v1.{Protos => Mesos}
+
+trait TaskBuilder {
```
Could you add docs since this is user facing?

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/template/SimpleRunTemplateFactory.scala:
```diff
@@ -0,0 +1,87 @@
+      dockerImageName: Option[String] = None)
+      extends TaskBuilder {
+
+    override def buildTask(
```
I wonder if we could/should make this private to `[usi]`. What do you think?

timcharper replied:
We cannot make it private; it's a publicly implementable trait method

ANeumann82 commented on core-models/src/main/scala/com/mesosphere/usi/core/linting/Linting.scala:
```diff
@@ -0,0 +1,47 @@
+    }
+  }
+
+  private def validateScalarRequirements(resourceRequirements: Seq[ResourceRequirement]): Seq[LintingWarning] = {
```
Nit: Maybe use the same signature as above, and pass in the full runTemplate instead of only the resourceRequirements?

Also: Should we validate that each scalar is only required once here, or is that done somewhere else? I.e. two requirements for "cpu" should be invalid

timcharper replied:
This is just moved code (out of Mesos Event Logic and in to something that can be optionally used if wanted). I'd rather not deal with it in detail right now.

ANeumann82 commented on core-models/src/main/scala/com/mesosphere/usi/core/models/template/RunTemplate.scala:
```diff
@@ -0,0 +1,129 @@
+}
+
+object RunTemplate extends StrictLogging {
+  case class KeyedResourceRequirement(entityKey: Option[TaskName], requirement: ResourceRequirement)
```
```suggestion
  /**
   * Used to identify to which task a requirement belongs. An empty entityKey indicates that the requirement belongs to the executor 
   **/
  case class KeyedResourceRequirement(entityKey: Option[TaskName], requirement: ResourceRequirement)
```

ANeumann82 commented on core-models/src/main/scala/com/mesosphere/usi/core/models/template/RunTemplate.scala:
```diff
@@ -0,0 +1,129 @@
+  final override val taskResourceRequirements: Map[TaskName, Seq[ResourceRequirement]] =
+    Map(TaskName.empty -> taskBuilder.resourceRequirements)
+
+  final override val executorResources: Seq[ResourceRequirement] = Nil
```
Shouldn't this be an empty seq? In the code above we access with with .foreach without checking for Nil...

```suggestion
  final override val executorResources: Seq[ResourceRequirement] = Seq.empty
```

timcharper replied:
Nil is List.empty which implements Seq

ANeumann82 commented on core/src/main/scala/com/mesosphere/usi/core/logic/MesosEventsLogic.scala:
```diff
@@ -19,25 +18,38 @@ import scala.collection.JavaConverters._
 private[core] class MesosEventsLogic(mesosCallFactory: MesosCalls, offerMatcher: OfferMatcher = new FCFSOfferMatcher())
     extends ImplicitStrictLogging {
 
+  val podTaskIdStrategy: PodTaskIdStrategy = PodTaskIdStrategy.DefaultStrategy
```
This will be settable later on, so we can use different strategies for e.g. legacy metronome IDs, right?

timcharper replied:
Yeah, that's the idea. Not all frameworks will have the same opinion about how task names in a taskGroup related to a taskId.

ANeumann82 commented on core/src/main/scala/com/mesosphere/usi/core/logic/MesosEventsLogic.scala:
```diff
@@ -121,28 +114,34 @@ private[core] class MesosEventsLogic(mesosCallFactory: MesosCalls, offerMatcher:
+                // http://mesos.apache.org/documentation/latest/scheduler-http-api/#acknowledge
+                List(mesosCallFactory.newAcknowledge(taskStatus.getAgentId, taskStatus.getTaskId, taskStatus.getUuid))
+              } else {
+                Nil
```
Do we need a warning or error log here? Or can we safely ignore it?

timcharper replied:
No, this is the expected way for frameworks to handle task status updates; acknowledge only needed if status has a uuid

ANeumann82 reviewed (approved):
Nothing serious, nice work

Commit:
TaskName containing period is fine

Commit:
Docs. Fix integration test. Change taskBuilder to receive a mutable builder object.

Commit:
Fix test

jeschkies committed:
Update core-models/src/main/scala/com/mesosphere/usi/core/models/template/RunTemplate.scala

Co-Authored-By: Andreas Neumann <aneumann@mesosphere.com>

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/template/RunTemplate.scala:
```diff
@@ -0,0 +1,133 @@
+      matchedOffer: Mesos.Offer,
+      taskName: TaskName,
+      taskIdStrategy: CurriedPodTaskIdStrategy,
+      resources: Seq[Mesos.Resource]): Unit = {
```
Are these the matched resources?

jeschkies committed:
Document simple builder.

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/logic/MesosEventsLogic.scala:
```diff
@@ -48,56 +60,37 @@ private[core] class MesosEventsLogic(mesosCallFactory: MesosCalls, offerMatcher:
-        LoggingArgs("offerId" -> offer.getId.getValue).and("mesosOperation" -> "LAUNCH")
-      )
+      val operations = launchCommands.values.map {
+        case Right(launchGroup) =>
```
Nice!

jeschkies committed:
Merge remote-tracking branch 'origin/master' into tharper/run-template

jeschkies committed:
Remove unused Proto builders.
